### PR TITLE
Added Reward Backpropogation Support 

### PR DESCRIPTION
**Pull request by mihirp1998**
Hi,

I have added support for AlignProp (https://align-prop.github.io/) for finetuning Stable Diffusion model using reward gradients. 

AlignProp directly backpropagate gradients from the reward model to the diffusion weights. Thus is about 25x more sample and compute efficient than policy gradient based methods like DDPO.

The current implementation seems to train effectively, almost within an hour on a single A100 while using Aesthetic reward model. Please find the attached loss and reward curves  + some qualitative results after training.

https://github.com/huggingface/diffusers/issues/7312

**Difference between DDPO and AlignProp:**

- DDPO uses PPO, which is a policy gradient method for aligning diffusion models.  AlignProp doesn't use policy gradients instead it directly backpropagates gradients from the reward function to diffusion denoising process, to maximize reward.

- AlignProp can only work when the reward function is differentiable, DDPO on other hand can handle non-differentiable reward functions, as it never backpropagates gradients from the reward function weights.

- As AlignProp takes benefit of the differentiability of the reward function as it backpropagates gradient. It is significantly more sample efficient than DDPO. 

- The loss function in AlignProp is simply the negative of the reward value outputed by the reward function, while in DDPO it's the PPO loss function.

- As the reward function is sitting on the RGB images. AlignProp requires to do the full denoising chain from Noise to RGB during training, while DDPO can instead sample random denoising timesteps, similar to diffusion training.

- DDPO and AlignProp both use LoRA and gradient checkpointing.

CC: @parthos86 @sayakpaul  @lvwerra @younesbelkada

W B Chart 4_25_2024, 3_02_39 AM
W B Chart 4_25_2024, 3_02_45 AM

Image Generations post training:
Screenshot 2024-04-25 at 3 02 30 AM

### CODE REVIEW

File: docs/source/alignprop_trainer.mdx [hunk 1]
```diff
@@ -0,0 +1,117 @@
+# Aligning Text-to-Image Diffusion Models with Reward Backpropagation
+
+## The why
+
+| Before | After finetuning |
+| --- | --- |
+| <div style="text-align: center"><img src="https://huggingface.co/datasets/trl-internal-testing/example-images/resolve/main/images/pre_squirrel.png"/></div> |  <div style="text-align: center"><img src="https://huggingface.co/datasets/trl-internal-testing/example-images/resolve/main/images/post_squirrel.png"/></div> |
+| <div style="text-align: center"><img src="https://huggingface.co/datasets/trl-internal-testing/example-images/resolve/main/images/pre_crab.png"/></div> |  <div style="text-align: center"><img src="https://huggingface.co/datasets/trl-internal-testing/example-images/resolve/main/images/post_crab.png"/></div> |
+|  <div style="text-align: center"><img src="https://huggingface.co/datasets/trl-internal-testing/example-images/resolve/main/images/pre_starfish.png"/></div> |  <div style="text-align: center"><img src="https://huggingface.co/datasets/trl-internal-testing/example-images/resolve/main/images/post_starfish.png"/></div> |
```
Comment (younesbelkada) [inline docs/source/alignprop_trainer.mdx:9]: These images are the ones generated from DDPO no?

File: docs/source/alignprop_trainer.mdx [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+
+## Getting started with Stable Diffusion finetuning with reinforcement learning
+
+The machinery for finetuning of Stable Diffusion models with reinforcement learning makes heavy use of HuggingFace's `diffusers`
+library. A reason for  stating this is that getting started requires a bit of familiarity with the `diffusers` library concepts, mainly two of them - pipelines and schedulers.
+Right out of the box (`diffusers` library), there isn't a `Pipeline` nor a `Scheduler` instance that is suitable for finetuning with reinforcement learning. Some adjustments need to made. 
+
+There is a pipeline interface that is provided by this library that is required to be implemented to be used with the `DDPOTrainer`, which is the main machinery for fine-tuning Stable Diffusion with reinforcement learning. **Note: Only the StableDiffusion architecture is supported at this point.**
```
Comment (younesbelkada) [inline docs/source/alignprop_trainer.mdx:18]: Here it references DDPO trainer

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot @mihirp1998 for your hardwork ! In principle this looks good ! 
I just have few questions with respect to the differences between this method and DDPO, could you clearly highlight either on the documentation or in this PR what are the major differences between DDPO and this algorithm ? 🙏 
I would also like to have a review from @sayakpaul if possible, what do you think of Reward Backpropagation ? 
Thanks !

File: docs/source/alignprop_trainer.mdx (hunk 1 above)
Comment (mihirp1998) [reply docs/source/alignprop_trainer.mdx:9]: Yes, i wanted to update them, although I wasn't sure how to do it, as they linked to a huggingface internal webpage https://huggingface.co/datasets/trl-internal-testing/

If you can guide me on how to do it, i can update them.

**Comment by mihirp1998** [comment]
> Thanks a lot @mihirp1998 for your hardwork ! In principle this looks good ! I just have few questions with respect to the differences between this method and DDPO, could you clearly highlight either on the documentation or in this PR what are the major differences between DDPO and this algorithm ? 🙏 I would also like to have a review from @sayakpaul if possible, what do you think of Reward Backpropagation ? Thanks !


I have added the differences in the pull request, let me know if u have some doubts or think something is missing.

File: docs/source/alignprop_trainer.mdx (hunk 2 above)
Comment (mihirp1998) [reply docs/source/alignprop_trainer.mdx]: Thanks for pointing this. I have fixed this.

File: docs/source/alignprop_trainer.mdx (hunk 1 above)
Comment (sayakpaul) [reply docs/source/alignprop_trainer.mdx:9]: You can open a PR to https://huggingface.co/datasets/trl-internal-testing/ repository adding the resultant images you want.

File: docs/source/alignprop_trainer.mdx [hunk 3]
```diff
@@ -0,0 +1,117 @@
+# Aligning Text-to-Image Diffusion Models with Reward Backpropagation
+
+## The why
```
Comment (sayakpaul) [inline docs/source/alignprop_trainer.mdx:3]: I don't think as a reader I understand if the following table justifies the name of this section. Would you mind elaborating?

File: docs/source/alignprop_trainer.mdx [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+| Before | After finetuning |
+| --- | --- |
+| <div style="text-align: center"><img src="https://huggingface.co/datasets/trl-internal-testing/example-images/resolve/main/images/pre_squirrel.png"/></div> |  <div style="text-align: center"><img src="https://huggingface.co/datasets/trl-internal-testing/example-images/resolve/main/images/post_squirrel.png"/></div> |
+| <div style="text-align: center"><img src="https://huggingface.co/datasets/trl-internal-testing/example-images/resolve/main/images/pre_crab.png"/></div> |  <div style="text-align: center"><img src="https://huggingface.co/datasets/trl-internal-testing/example-images/resolve/main/images/post_crab.png"/></div> |
+|  <div style="text-align: center"><img src="https://huggingface.co/datasets/trl-internal-testing/example-images/resolve/main/images/pre_starfish.png"/></div> |  <div style="text-align: center"><img src="https://huggingface.co/datasets/trl-internal-testing/example-images/resolve/main/images/post_starfish.png"/></div> |
+
+
+## Getting started with Stable Diffusion finetuning with reinforcement learning
```
Comment (sayakpaul) [inline docs/source/alignprop_trainer.mdx:12]: I don't think this is needed. We should strive to keep the API documentation lean and precise.

File: docs/source/alignprop_trainer.mdx [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+```
+
+### Using the finetuned model
+
+Assuming you've done with all the epochs and have pushed up your model to the hub, you can use the finetuned model as follows
+
+```python
+
+import torch
+from trl import DefaultDDPOStableDiffusionPipeline
```
Comment (sayakpaul) [inline docs/source/alignprop_trainer.mdx:95]: Why do we have use a non-diffusers pipeline here? Does `DiffusionPipeline` from `diffusers` not work here?

File: docs/source/alignprop_trainer.mdx [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+pipeline = DefaultDDPOStableDiffusionPipeline("metric-space/alignprop-finetuned-sd-model")
+
+device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
+
+# memory optimization
+pipeline.vae.to(device, torch.float16)
+pipeline.text_encoder.to(device, torch.float16)
+pipeline.unet.to(device, torch.float16)
```
Comment (sayakpaul) [inline docs/source/alignprop_trainer.mdx:104]: These LoCs could be reduce if we do:

```python
pipeline = DefaultDDPOStableDiffusionPipeline("metric-space/alignprop-finetuned-sd-model", torch_dtype=torch.float16)
pipeline = pipeline.to("cuda")
```

Comment (sayakpaul) [reply docs/source/alignprop_trainer.mdx:104]: Additionally, https://huggingface.co/metric-space/alignprop-finetuned-sd-model is not available. Let's make sure we're using the right checkpoint ids here.

File: examples/scripts/alignprop.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+    def forward(self, embed):
+        return self.layers(embed)
+
+
+class AestheticScorer(torch.nn.Module):
+    """
+    This model attempts to predict the aesthetic score of an image. The aesthetic score
+    is a numerical approximation of how much a specific image is liked by humans on average.
+    This is from https://github.com/christophschuhmann/improved-aesthetic-predictor
```
Comment (sayakpaul) [inline examples/scripts/alignprop.py:84]: Why are we copy-pasting these modules from the DDPO script? 

@younesbelkada would it make sense to have a separate module for these (`auxiliary_modules`, perhaps)?

File: trl/models/modeling_sd_base.py [hunk 8]
```diff
@@ -527,6 +527,243 @@ def pipeline_step(
...
+
+
+            if truncated_backprop:
+                if truncated_backprop_rand:
+                    rand_timestep = random.randint(truncated_rand_backprop_minmax[0],truncated_rand_backprop_minmax[1])
+                    if i < rand_timestep:
+                        noise_pred = noise_pred.detach()
+                else:
+                    if i < truncated_backprop_timestep:
+                        noise_pred = noise_pred.detach()
```
Comment (sayakpaul) [inline trl/models/modeling_sd_base.py:722]: We would want to supplement this code block with comments.

File: trl/models/modeling_sd_base.py [hunk 9]
```diff
@@ -527,6 +527,243 @@ def pipeline_step(
 
     return DDPOPipelineOutput(image, all_latents, all_log_probs)
 
+def pipeline_step_with_grad(
+    self,
```
Comment (sayakpaul) [inline trl/models/modeling_sd_base.py:531]: `self`, could be replace with `pipeline` as that is what we're passing down the line, IIUC?

File: trl/trainer/alignprop_trainer.py [hunk 10]
```diff
@@ -0,0 +1,439 @@
...
+- alignprop
+- diffusers
+- reinforcement-learning
+- text-to-image
+- stable-diffusion
+---
+
+# {model_name}
+
+This is a pipeline that finetunes a diffusion model with reward gradients. The model can be used for image generation conditioned with text.
```
Comment (sayakpaul) [inline trl/trainer/alignprop_trainer.py:50]: Not sure what is the norm is within the library but I think it could be nice to also include a link to the AlignProp paper here.

**Review by sayakpaul — COMMENTED** [review]
Thanks for your contributions. I left a couple of comments. 

I would love to see some concrete comparisons to DDPO (training time, reward dynamics, convergence of the validation samples, etc.).

**Comment by mihirp1998** [comment]
> Thanks for your contributions. I left a couple of comments.
> 
> I would love to see some concrete comparisons to DDPO (training time, reward dynamics, convergence of the validation samples, etc.).


I have made concrete comparisions with DDPO here. I ran the DDPO default code in TRL with batch size 128, while AlignProp also uses the same batch size. As can be seen AlignProp is significantly more sample efficient, here i train both the models for a few hours. Here x-axis is the epochs and y-axis is the reward achieved.


W B Chart 6_1_2024, 10_35_51 PM (1)

Although i ran the above experiments for a few hours, AlignProp only takes about 30 minutes to converge to a good solution. So i early stopped at the 8th epoch in training. Below is the comparision with DDPO after training both models for 30 minutes. Here x-axis is the training time and y-axis is the reward achieved.

W B Chart 6_1_2024, 10_35_51 PM

Both the curves are similar to the curves in the paper.

The above curves were with the same set of prompts during training/testing. In the curve below i show AlignProp results on unseen prompts. As can be seen there is not much gap in results between seen/unseen prompts. Here dotted lines is the unseen prompts while solid line is the seen prompts.


W B Chart 6_1_2024, 10_34_58 PM

Finally here are some generated images from AlignProp, for seen/unseen animals after training.

media_images_llama_8_d86f465673c821a39d26 (1)
media_images_lion_8_9e9a947367e031acdceb (1)
squirrel

CC: @sayakpaul @younesbelkada

File: trl/trainer/alignprop_trainer.py (hunk 10 above)
Comment (mihirp1998) [reply trl/trainer/alignprop_trainer.py]: Yes i added.

File: trl/models/modeling_sd_base.py (hunk 9 above)
Comment (mihirp1998) [reply trl/models/modeling_sd_base.py]: Yes i changed it to pipeline.

File: trl/models/modeling_sd_base.py (hunk 8 above)
Comment (mihirp1998) [reply trl/models/modeling_sd_base.py]: Yes added comments.

File: docs/source/alignprop_trainer.mdx (hunk 6 above)
Comment (mihirp1998) [reply docs/source/alignprop_trainer.mdx]: Yes i reduced it and fixed the checkpoint ids.

File: docs/source/alignprop_trainer.mdx (hunk 5 above)
Comment (mihirp1998) [reply docs/source/alignprop_trainer.mdx]: Yes indeed, i changed it to StableDiffusionPipeline from diffusers

File: docs/source/alignprop_trainer.mdx (hunk 4 above)
Comment (mihirp1998) [reply docs/source/alignprop_trainer.mdx]: Yes i removed it.

File: docs/source/alignprop_trainer.mdx (hunk 3 above)
Comment (mihirp1998) [reply docs/source/alignprop_trainer.mdx:3]: Thanks,  I added a better why statement.

File: examples/scripts/alignprop.py (hunk 7 above)
Comment (mihirp1998) [reply examples/scripts/alignprop.py]: They are not exactly copy pasted, as DDPO had clamp and no_grad operations within them, which were preventing gradients from backpropagating.

Anyhow I still transfered the above reward function code from alignprop.py to [trl/models/auxiliary_modules.py](https://github.com/huggingface/trl/pull/1585/commits/c0a6ce32014db391e19b24e36123bd4fcb258c8c#diff-02816a76d7dad2e2a0e985ba808ee10137c85eddf18ea6df322cb079ba712387), as u suggested.

**Comment by mihirp1998** [comment]
Hi @sayakpaul @younesbelkada ,

Any update on this? Happy to fix things if something's missing.

File: trl/__init__.py [hunk 11]
```diff
@@ -4,6 +4,7 @@
 
 from typing import TYPE_CHECKING
 from .import_utils import _LazyModule, is_diffusers_available, OptionalDependencyNotAvailable
+import torch.utils.checkpoint as checkpoint
```
Comment (younesbelkada) [inline trl/__init__.py:7]:
```suggestion
```
This is not needed no?

File: docs/source/alignprop_trainer.mdx
Comment (younesbelkada) [inline docs/source/alignprop_trainer.mdx:1]: Very nice documentation page !

**Review by younesbelkada — COMMENTED** [review]
Looks clean on my end ! Just one comment about an import that is not needed and we should be good to go !

File: docs/source/alignprop_trainer.mdx
Comment (mihirp1998) [reply docs/source/alignprop_trainer.mdx:1]: Thanks! Yes it wasn't needed, i committed it.

**Review by younesbelkada — APPROVED** [review]
Very nice integration thanks for adding this !

**Comment by younesbelkada** [comment]
cc @sayakpaul could you give a final look ?

**Comment by younesbelkada** [comment]
@mihirp1998 can you run the styling checks? `make precommit`

**Comment by sayakpaul** [comment]
Will do. Allow me today as today is the SD3 release :)

**Comment by younesbelkada** [comment]
Sure, no rush

**Comment by younesbelkada** [comment]
@mihirp1998 can you also add `alignprop_trainer` in the documentation `_toctree.yaml` file?

**Review by sayakpaul — APPROVED** [review]
Great work. I really like the sample efficiency gains :)

**Comment by mihirp1998** [comment]
> Great work. I really like the sample efficiency gains :)

Thanks! Is something left from my side? As it still says: **2 workflows awaiting approval**

**Comment by mihirp1998** [comment]
> > Great work. I really like the sample efficiency gains :)
> 
> Thanks! Is something left from my side? As it still says: **2 workflows awaiting approval**

Oh nvm, seems like everything is approved! Thanks!

**Comment by younesbelkada** [comment]
cc @vwxyzjn for a final look, if all looks great to you, feel free to merge!

**Comment by mihirp1998** [comment]
Hi @vwxyzjn ,

Any update on this? Happy to fix things in case something's missing.

**Review by vwxyzjn — APPROVED** [review]
Looks great! Thanks for the work!
